language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -39,14 +48,42 @@ data TypeConfig = TypeConfig
     { _typeConfigNormalize           :: Bool
       -- ^ Whether to normalize type annotations.
     , _typeConfigDynBuiltinNameTypes :: DynamicBuiltinNameTypes
-    , _typeConfigGas                 :: GasInit -- ^ The upper limit on the length of type
-                                      -- reductions. If set to 'Nothing', type
-                                      -- reductions will be unbounded.
+    , _typeConfigGas                 :: Maybe Gas
+       -- ^ The upper limit on the length of type reductions.
+       -- If set to 'Nothing', type reductions will be unbounded.
     }
 
 -- | The type checking monad contains the 'BuiltinTable' and it lets us throw
 -- 'TypeError's.
-type TypeCheckM a = ReaderT TypeConfig (ExceptT (TypeError a) Quote)
+type TypeCheckM ann = ReaderT TypeConfig (ExceptT (TypeError ann) Quote)
+
+-- | Run a 'NormalizeTypeT' computation in the 'TypeCheckM' context.
+-- Depending on whether gas is finite or infinite, calls either
+-- 'runNormalizeTypeAnyM' or 'runNormalizeTypeGasM'.
+-- Throws 'OutOfGas' if type normalization runs out of gas.
+runInTypeCheckM :: (forall m. MonadQuote m => NormalizeTypeT m tyname ann1 a) -> TypeCheckM ann2 a
+runInTypeCheckM a = do
+    mayGas <- _typeConfigGas <$> ask
+    case mayGas of
+        Nothing  -> runNormalizeTypeAnyM a
+        Just gas -> do
+            mayX <- liftQuote $ runNormalizeTypeGasM gas a
+            case mayX of
+                Nothing -> throwing _TypeError OutOfGas
+                Just x  -> pure x
+
+-- | Normalize a 'Type'.
+normalizeTypeTcm :: Type TyNameWithKind () -> TypeCheckM a (NormalizedType TyNameWithKind ())
```

Comment:
> What does Tcm mean?

`Tcm` is an abbreviation of `TypeCheckM` (people often give their type checking monad the `TCM` name). I wanted to stress that this function runs in `TypeCheckM` and didn't want to write `normalizeTypeTypeCheckM`, because that's terrible even for me.

> Also, can't this now be generic over annotations too?

It can, but the function is only used locally (it's not exported) and over types annotated by `()`, so I prefer to be restrictive here.